src/Language/DifferentialDatalog/CompileJava.hs

```diff
@@ -0,0 +1,487 @@
+import Language.DifferentialDatalog.DatalogProgram
+import Language.DifferentialDatalog.Relation
+import Language.DifferentialDatalog.Optimize
+import Language.DifferentialDatalog.Module
+import Language.DifferentialDatalog.ECtx
+import Language.DifferentialDatalog.Type
+import Language.DifferentialDatalog.Rule
+
+-- generate a Java program from a DDlog program
+-- The Java program has some utility function to more easily create
+-- records for the DDlog relations.  Note that Java does not support
+-- all DDlog constructs
+compileJava :: DatalogProgram -> String -> IO ()
+compileJava d sourceName = do
+    let content = generateJava d sourceName in
+         writeFile (capitalize sourceName ++ ".java") (render content)
+
+-- Generate a Java document from a program given the source filename
+generateJava :: DatalogProgram -> String -> Doc
+generateJava d sourceName =
+    preamble sourceName $+$
+    (braces' $ body d)
+
+-- Generates the preamble of a Java file
+preamble :: String -> Doc
+preamble sourceName =
+    "import java.util.*;"                       $+$
+    "import ddlogapi.DDlogRecord;\n"            $+$
+    "import ddlogapi.DDlogCommand;\n"           $+$
+    "class" <+> pp (capitalize sourceName)
+
+data JavaState = JavaState {
+    -- types that have been compiled
+    typesCompiled  :: S.Set String,
+    -- types that still have to be compiled
+    typesToCompile :: M.Map String Type
+}
+
+type JavaMonad = State JavaState
+emptyState :: JavaState
+emptyState = JavaState {
+    typesCompiled  = S.empty,
+    typesToCompile = M.empty
+}
+
+-- Generates the body of the Java class containing all utility methods
+body :: DatalogProgram -> Doc
+body d =
+    evalState (
+        do relations <- mapM (genRelationCreate d) (M.keys $ progRelations d)
+           -- TODO: this should iterate until todo is empty
+           todo <- gets (\state -> M.toList $ typesToCompile state)
+           creates <- mapM (\(n,t) -> genCreate d n t) todo
+           let types = map (\(n,t) -> genClass d n t False) todo
+           return $ genEnum d $+$
+                 (vcat $ punctuate "\n" relations) $+$
+                 (vcat $ punctuate "\n" $ map (genRelationClass d) (M.keys $ progRelations d)) $+$
+                 (vcat $ punctuate "\n" types) $+$
+                 (vcat $ punctuate "\n" creates)
+    )
+    emptyState
+
+-- Generate a Java class for a specific relation
+genRelationClass :: DatalogProgram -> String -> Doc
+genRelationClass d@DatalogProgram{..} relationName =
+    let rtype = relType $ progRelations M.! relationName in
+    genClass d relationName rtype True
+
+-- Generate a Java class for a specific type
+genClass :: DatalogProgram -> String -> Type -> Bool -> Doc
+genClass d@DatalogProgram{} name rtype isRelation =
+    "public static class" <+> makeIdentifier name $+$
+    (braces' $ classBody d name rtype isRelation)
+
+-- given a type generate the body of a Java class storing the columns of the relation
```

Comment:
Isn't supporting multiple constructors just a matter of generating a class per constructor?

Reply:
That's one way to do it. These classes would also probably need to inherit from a common interface. Let's leave this until someone needs this feature.